# Taller 00 - Analítica de Datos
- **Autor:** David Santiago Barreto Mora.
- **Fecha de entrega:** 10/08/2023 
- **Profesor:** Ing. John Corredo, PhD.
---

## Fuente de datos
Para el desarrollo del taller, se usan los archivos de extension .csv dados por el profesor. Los archivos detallan las ventas hechas en Colombia en 2019, de lo que se presume es una tienda de tecnología. En total son 12 archivos, cada uno correspondiendo a las ventas de un mes, por lo que el dataset en total constituye las ventas anuales del 2019 de la empresa en cuestión.

Para poder comenzar realizar el trabajo analítico, primero importamos las librerias a ser usadas, y por supuesto, los archivos .csv.



### Carga de librerias


In [3]:
# Se importan las librerias que serán usadas en el análisis.
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt


### Carga de los datos (archivos .csv) 
A continuación se cargan los 12 archivos CSV y se crea un dataframe por cada uno. Esto con la intención de unir todos los dataframes en uno solo, para posteriormente facilitar su manipulación y análisis.

In [4]:
# Se importan los 12 archivos .csv, y por cada uno se crea un dataframe (libreria pandas)
dfJanuary = pd.read_csv("Ventas/Ventas_January_2019.csv")
dfFebruary = pd.read_csv("Ventas/Ventas_February_2019.csv")
dfMarch = pd.read_csv("Ventas/Ventas_March_2019.csv")
dfApril = pd.read_csv("Ventas/Ventas_April_2019.csv")
dfMay = pd.read_csv("Ventas/Ventas_May_2019.csv")
dfJune = pd.read_csv("Ventas/Ventas_June_2019.csv")
dfJuly = pd.read_csv("Ventas/Ventas_July_2019.csv")
dfAugust = pd.read_csv("Ventas/Ventas_August_2019.csv")
dfSeptember = pd.read_csv("Ventas/Ventas_September_2019.csv")
dfOctober = pd.read_csv("Ventas/Ventas_October_2019.csv")
dfNovember = pd.read_csv("Ventas/Ventas_November_2019.csv")
dfDecember = pd.read_csv("Ventas/Ventas_December_2019.csv")


A continuación imprimimos las 5 primeras fila de cada dataframe. Principalmente, para ver que las columnas de todos los dataframes sean las mismas. Esta verificicación es hecha para validar que se puedan unir los dataframes en uno solo.

In [5]:
# Se muestra la primera columna del dataframe de enero.
dfJanuary.head(5)


ID Orden                 Producto   Cantidad Ordenada Precio Unitario   \
0   141234   Audifonos Apple Airpods                  1           499900   
1   141235  Parlante Google HomeMini                  1           159900   
2   141236  Cable de Carga Lightning                  1            64900   
3   141237  Parlante Google HomeMini                  1           159900   
4   141238       Samsung Galaxy A302                  1           649900   

            Fecha              Dirección de Compra  
0  01/25/19 13:51      177 calle Jimenez, Cali, VC  
1  01/12/19 23:58        732 calle 9na, Cúcuta, NS  
2  01/15/19 21:42  887 calle Kennedy, Medellin, AN  
3  01/27/19 11:39   641 calle 5ta, Bucaramanga, SA  
4  01/14/19 17:13      379 calle 2da, Medellin, AN

In [6]:
# Se muestra la primera columna del dataframe de febrero.
dfFebruary.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   153258      Multiolla Oster 0150 Negra PE                  1   
1   153259              AA Batteries (4-pack)                  2   
2   153260  Horno Microondas 0.8P Samsung AGE                  1   
3   153261           Cable de Carga Lightning                  1   
4   153262           Parlante Google HomeMini                  1   

  Precio Unitario            Fecha              Dirección de Compra  
0           299900  02/04/19 07:28      724 calle Adams, Cúcuta, NS  
1            13800  02/22/19 20:19      61 calle Pinar, Armenia, QU  
2           309900  02/12/19 12:34   40 calle Jimenez, Medellin, AN  
3            64900  02/09/19 18:07  670 calle Jefferson, Bogotá, CU  
4           159900  02/05/19 18:34    223 calle Boyacá, Armenia, QU

In [7]:
# Se muestra la primera columna del dataframe de marzo.
dfMarch.head(5)


ID Orden                  Producto   Cantidad Ordenada Precio Unitario   \
0   163171  LG Secadora Carga Frontal                  1          1646900   
1   163172      AA Batteries (4-pack)                  1            13800   
2   163173  Audifonos Bose SoundSport                  1           399900   
3   163174   Cable de Carga USB-C P/O                  2            46900   
4   163175   Cable de Carga Lightning                  1            64900   

            Fecha                      Dirección de Compra  
0  03/31/19 11:58            591 calle Adams, Medellin, AN  
1  03/23/19 13:24             106 calle Adams, Armenia, QU  
2  03/08/19 13:46            545 calle Rivera, Armenia, QU  
3  03/30/19 18:04  370 calle Autopista Norte, Medellin, AN  
4  03/06/19 21:08              347 calle Sucre, Ibagué, TO

In [8]:
# Se muestra la primera columna del dataframe de abril
dfApril.head(5)


ID Orden                 Producto   Cantidad Ordenada Precio Unitario   \
0   171346  Cable de Carga USB-C P/O                  1            46900   
1   171347  Parlante Google HomeMini                  1           159900   
2   171348    AAA Batteries (4-pack)                  1            12990   
3   171349  Cable de Carga USB-C P/O                  1            46900   
4   171350  Cable de Carga USB-C P/O                  1            46900   

            Fecha            Dirección de Compra  
0  04/04/19 19:25  631 calle Boyacá, Armenia, QU  
1  04/19/19 07:19    546 calle 5ta, Medellin, AN  
2  04/22/19 22:41   593 calle 8va, Manizalez, CA  
3  04/09/19 04:59    571 calle 4ta, Medellin, AN  
4  04/18/19 09:36   883 calle Madison, Neiva, HU

In [9]:
# Se muestra la primera columna del dataframe de mayo.
dfMay.head(5)


ID Orden                  Producto   Cantidad Ordenada Precio Unitario   \
0   182565  LG Lavadora 19Kg Inv WT19                  1          1646900   
1   182566        Xiaomi REDMI Note 8                  1           719000   
2   182567      AA Batteries (4-pack)                  1            13800   
3   182568   Cable de Carga Lightning                  1            64900   
4   182569   Parlante Google HomeMini                  1           159900   

            Fecha            Dirección de Compra  
0  05/05/19 10:02    722 calle Adams, Bogotá, CU  
1  05/04/19 22:44  986 calle Johnson, Bogotá, CU  
2  05/02/19 17:24     422 calle Rio, Armenia, QU  
3  05/25/19 22:49      630 calle 7ma, Bogotá, CU  
4  05/20/19 17:24   741 calle 7ma, Manizalez, CA

In [10]:
# Se muestra la primera columna del dataframe de junio.
dfJune.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   195098           Cable de Carga Lightning                  1   
1   195099           Cable de Carga Lightning                  1   
2   195100  Freidora Air Fryer Kalley K-MAF25                  1   
3   195101             AAA Batteries (4-pack)                  1   
4   195102              AA Batteries (4-pack)                  1   

  Precio Unitario            Fecha                  Dirección de Compra  
0            64900  06/17/19 20:12  736 calle Transversal, Medellin, AN  
1            64900  06/25/19 00:27          562 calle Elm, Medellin, AN  
2           198849  06/23/19 20:44        548 calle Sucre, Medellin, AN  
3            12990  06/12/19 20:30          322 calle Sur, Medellin, AN  
4            13800  06/02/19 10:00        928 calle Avenida, Ibagué, TO

In [11]:
# Se muestra la primera columna del dataframe de julio.
dfJuly.head(5)


ID Orden                       Producto   Cantidad Ordenada  \
0   211128          AAA Batteries (4-pack)                  1   
1   211129  Asus VivoBook X420FA-BV Laptop                  1   
2   211130   Multiolla Oster 0150 Negra PE                  1   
3   211131           AA Batteries (4-pack)                  1   
4   211132           AA Batteries (4-pack)                  1   

  Precio Unitario            Fecha                 Dirección de Compra  
0            12990  07/09/19 21:48       327 calle Carrera, Cúcuta, NS  
1          1950900  07/14/19 15:56        597 calle Rio, Manizalez, CA  
2           299900  07/28/19 21:25          658 calle 14va, Ibagué, TO  
3            13800  07/24/19 12:47  100 calle Transversal, Armenia, QU  
4            13800  07/20/19 18:30         494 calle Elm, Medellin, AN

In [12]:
# Se muestra la primera columna del dataframe de agosto.
dfAugust.head(5)


ID Orden                 Producto   Cantidad Ordenada Precio Unitario   \
0   224782    AAA Batteries (4-pack)                  2            12990   
1   224783    AAA Batteries (4-pack)                  1            12990   
2   224784     AA Batteries (4-pack)                  1            13800   
3   224785  Cable de Carga Lightning                  1            64900   
4   224786    AAA Batteries (4-pack)                  1            12990   

            Fecha             Dirección de Compra  
0  08/04/19 08:24    91 calle Boyacá, Armenia, QU  
1  08/31/19 07:55  157 calle Willow, Medellin, AN  
2  08/30/19 18:25   317 calle Santa, Medellin, AN  
3  08/13/19 13:48      744 calle Centro, Cali, VC  
4  08/08/19 10:52   486 calle Cherry, Armenia, QU

In [13]:
# Se muestra la primera columna del dataframe de septiembre.
dfSeptember.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   242703                Samsung Galaxy A302                  1   
1   242704           Cable de Carga Lightning                  1   
2   242705          Audifonos Bose SoundSport                  1   
3   242706  Horno Microondas 0.8P Samsung AGE                  1   
4   242707  Freidora Saludable B&D 3.5 Litros                  1   

  Precio Unitario            Fecha                  Dirección de Compra  
0           649900  09/19/19 12:57        508 calle 11ra, Manizalez, CA  
1            64900  09/10/19 20:36          401 calle 9na, Medellin, AN  
2           399900  09/24/19 18:44  320 calle Portal Sur, Manizalez, CA  
3           309900  09/23/19 12:19      760 calle 11ra, Bucaramanga, SA  
4           229900  09/03/19 00:09        5 calle Diagonal, Armenia, QU

In [14]:
# Se muestra la primera columna del dataframe de octubre.
dfOctober.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   246400  Freidora Saludable B&D 3.5 Litros                  1   
1   246401           Parlante Google HomeMini                  1   
2   246402              AA Batteries (4-pack)                  1   
3   246403           Cable de Carga USB-C P/O                  1   
4   246404           Cable de Carga USB-C P/O                  1   

  Precio Unitario            Fecha                   Dirección de Compra  
0           229900  10/05/19 17:01         893 calle 14va, Manizalez, CA  
1           159900  10/09/19 13:20         502 calle Chapinero, Cali, VC  
2            13800  10/19/19 13:04          202 calle Cedar, Armenia, QU  
3            46900  10/21/19 03:12  954 calle Americas, Barranquilla, AT  
4            46900  10/10/19 21:15           79 calle Sunset, Ibagué, TO

In [15]:
# Se muestra la primera columna del dataframe de noviembre.
dfNovember.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   260944  Freidora Saludable B&D 3.5 Litros                  1   
1   260945            Audifonos Apple Airpods                  1   
2   260946           Cable de Carga USB-C P/O                  2   
3   260947          Audifonos Bose SoundSport                  1   
4   260948            Audifonos Apple Airpods                  1   

  Precio Unitario            Fecha                     Dirección de Compra  
0           229900  11/12/19 14:57           79 calle Avenida, Armenia, QU  
1           499900  11/04/19 17:52           849 calle Atalaya, Bogotá, CU  
2            46900  11/20/19 11:37                99 calle 9na, Ibagué, TO  
3           399900  11/26/19 17:58  126 calle Portal Sur, Barranquilla, AT  
4           499900  11/27/19 18:16           983 calle Museo, Medellin, AN

In [16]:
# Se muestra la primera columna del dataframe de diciembre.
dfDecember.head(5)


ID Orden                      Producto   Cantidad Ordenada Precio Unitario   \
0   276339      Audifonos Bose SoundSport                  1           399900   
1   276340  Multiolla Oster 0150 Negra PE                  1           299900   
2   276341            Samsung Galaxy A302                  1           649900   
3   276342       Cable de Carga Lightning                  1            64900   
4   276343      Audifonos Bose SoundSport                  1           399900   

            Fecha                Dirección de Compra  
0  12/11/19 19:29    345 calle Pte Aranda, Neiva, HU  
1  12/13/19 16:56  924 calle Americas, Manizalez, CA  
2  12/12/19 06:05     616 calle Sunset, Medellin, AN  
3  12/23/19 21:22     189 calle Willow, Medellin, AN  
4  12/20/19 15:36      931 calle Madison, Bogotá, CU

Como se puede ver, todos los dataframes (o archivos .csv) cuentan con 6 columnas:
1. ID Orden
2. Producto
3. Cantidad Ordenada
4. Precio Unitario
5. Fecha
6. Dirección de Compra

### Unión de los dataframes 
Para poder consolidar todos los datos en un solo archivo, es necesario entender como funciona la función merge (ya que es requerido unir los dataframes con esta función). 

Si nos guiamos por la documentación oficial, la función es casi que análoga a un JOIN de SQL. Esto es supremamente importante de entender, porque la tabla resultante de un JOIN puede tener nuevas filas o columnas que no estaban previamente en ninguna de las 2 tablas usadas. Lo que nosotros buscamos en este análisis es anexar los datos de una tabla "debajo" de otra, con el fin de poder tener toda la información de los 12 meses sin alteración alguna y de manera continua. lo cual es problemático teniendo en cuenta las nuevas filas que puede generar un join. 

A continuación una demostración del problema que queremos evitar. Supongamos que vamos a usar 'merge()' con las tablas de Enero y Febrero. Primero mostramos la cantidad de filas y columnas de cada dataframe.

In [17]:
print("Dimensiones del Dataframe de Enero:", dfJanuary.shape)
print("Dimensiones del Dataframe de Enero:", dfFebruary.shape)

Dimensiones del Dataframe de Enero: (12554, 6)
Dimensiones del Dataframe de Enero: (10355, 6)


Podemos ver que las tablas tienen 12554 y 10355 filas, respectivamente. Por ahora solo nos interesa el numero de filas.

Si ahora ejecutamos la función merge con los dos dataframes:

In [19]:
mergedDf = pd.merge(dfJanuary, dfFebruary, how="outer")
print("Dimensiones del Dataframe de Enero:", mergedDf.shape)
print(12554 + 10355)

Dimensiones del Dataframe de Enero: (24474, 6)
22909


Podemos ver que el dataframe combinado tiene en total 24474 filas, cuando la suma de las filas de las tablas originales da 22909. Más de 1000 registros nuevos fueron creados. También resalto el parámetro "how", el cual especifica como unir las tablas (inner join, outer join, left outer join, etc). El ejemplo se muestra con "outer" porque esto es lo más parecido a unir totalmente las 2 tablas. 

In [20]:
dfJanuary['dummy'] = 0
dfFebruary['dummy'] = 0

mergedDf = pd.merge(dfJanuary, dfFebruary, how="outer", on='dummy')
print(mergedDf.shape)


(129996670, 13)
